[14.0][MIG] sale_blanket_order: Migration to 14.0

janikvonrotz opened the pull request:
Standard migration from 13.0 to 14.0.

andreparames committed:
[10.0][ADD] sale_blanket_orders

andreparames committed:
sale_blanket_order: don't copy name nor confirm state

When duplicating a confirmed blanket order, the new copy shouldn't
keep the state nor the sequence number (name).

andreparames committed:
sale_blanket_order: update after review

andreparames committed:
sale_blanket_order: Re-calculate price based on qty

andreparames committed:
sale_blanket_order: make prohibition of adding new lines optional

rousseldenis committed:
[FIX] sale_blanket_order: Ondelete cascade on blanket lines

AdriaGForgeFlow committed:
[11.0][MIG] Migrate module sale_blanket_order

OCA-git-bot committed:
[ADD] icon.png

OCA-git-bot committed:
[UPD] README.rst

oca-travis committed:
[UPD] Update sale_blanket_order.pot

marylla committed:
Added translation using Weblate (German)

marylla committed:
Translated using Weblate (German)

Currently translated at 7.3% (9 of 123 strings)

Translation: sale-workflow-11.0/sale-workflow-11.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-11-0/sale-workflow-11-0-sale_blanket_order/de/

OCA-git-bot committed:
[UPD] README.rst

tafaRU committed:
[MIG] sale_blanket_order: Migration to 12.0

tafaRU committed:
sale_blanket_order: fix the following use case scenario.

Steps to reproduce:

* create and confirm a blanket order (BO) with (product A, qty 30) and (product B, qty 20)
* from the BO create a SO with (product A, qty 10) and (product B, qty 10)
* from the BO create a SO with (product A, qty 20) and (product B, qty 0)
* from the BO create another SO with (product B, qty 10)

Current behavior:

It raises the exception "The sale has already been completed.".

Expected behavior:

No exception is raised.

tafaRU committed:
sale_blanket_order: correctly display images in both README and index

oca-travis committed:
[UPD] Update sale_blanket_order.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_blanket_order/

tafaRU committed:
sale_blanket_order: fix my email address format in CONTRIBUTORS.rst

Galoperd committed:
Added translation using Weblate (Spanish)

Galoperd committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (154 of 154 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_blanket_order/es/

pedrocs-exo committed:
Added translation using Weblate (Portuguese)

AndreVolksdorf committed:
Translated using Weblate (German)

Currently translated at 7.1% (11 of 154 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_blanket_order/de/

AndreVolksdorf committed:
Translated using Weblate (German)

Currently translated at 53.9% (83 of 154 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_blanket_order/de/

AndreVolksdorf committed:
Translated using Weblate (German)

Currently translated at 100.0% (154 of 154 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_blanket_order/de/

AndreVolksdorf committed:
Translated using Weblate (German)

Currently translated at 100.0% (154 of 154 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_blanket_order/de/

chusamo committed:
[MIG]sale_blanker_order: Migration to 13.0

oca-travis committed:
[UPD] Update sale_blanket_order.pot

OCA-git-bot committed:
[UPD] README.rst

LoisRForgeFlow committed:
[IMP] sale_blanket_order: use default terms and conditions.

exactly like in sales orders.

LoisRForgeFlow committed:
[FIX] sale_blanket_order: multi-company rules were not migrated properly.

LoisRForgeFlow committed:
[IMP] sale_blanket_order: analytic acc and origin

* Allow to select an analitic account in the blanket order. It
  will be passed on to the SO.
* Allow to select analytic tags on blanket order lines. They will
  be passed to the SO lines.
* Source document was not being filled in SO created from a
  blanket order.

oca-travis committed:
[UPD] Update sale_blanket_order.pot

OCA-git-bot committed:
sale_blanket_order 13.0.1.1.0

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-13.0/sale-workflow-13.0-sale_blanket_order
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-13-0/sale-workflow-13-0-sale_blanket_order/

oca-travis committed:
[UPD] Update sale_blanket_order.pot

janikvonrotz committed:
[IMP] sale_blanket_order: black, isort, prettier

janikvonrotz committed:
[MIG] sale_blanket_order: Migration to 14.0

janikvonrotz commented on sale_blanket_order/wizard/create_sale_orders.py:
```diff
@@ -0,0 +1,201 @@
+        pricelist_id = 0
+        user_id = 0
+        payment_term_id = 0
+        for line in self.line_ids.filtered(lambda l: l.qty != 0.0):
```
I am getting empty lines here.

hieutran-novobi replied:
Hi, I'm migrating this to 15 and I got the line.remaining_uom_qty equals zero to here. Any ideas how this could happen?

janikvonrotz committed:
[IMP] sale_blanket_order: tags
* Add tags and pass them to sale order

janikvonrotz committed:
[IMP] sale_blanket_order: invoice and shipping partner
* Define invoice and shipping address similar to sale order
* Pass addresses to sale order

janikvonrotz committed:
[IMP] sale_blanket_order: cancel state
* Add cancel state similar to sale order
* Set cancel state on cancel action

janikvonrotz committed:
[IMP] sale_blanket_order: confirmation date
* Set confirmation date when blanket order is confirmed

janikvonrotz committed:
[FIX] sale_blanket_order: bump minor version

janikvonrotz committed:
[FIX] sale_blanket_order: do not use supplier tax

janikvonrotz commented:
@LoisRForgeFlow I think you used the wrong tax (use sale not purchase). I fixed it this with: https://github.com/OCA/sale-workflow/pull/1785/commits/38906e1ed5113a57c6b0c7f8718f36cd83e1823c

LoisRForgeFlow commented:
> @LoisRForgeFlow I think you used the wrong tax (use sale not purchase). I fixed it this with: [38906e1](https://github.com/OCA/sale-workflow/commit/38906e1ed5113a57c6b0c7f8718f36cd83e1823c)

Seems that you are right, but just wanted to clarify that it wasn't me :) (https://github.com/OCA/sale-workflow/commit/7006e8a2e763a402142dfc8000b0604f698d5c5d#diff-cd64445e9587af7eb87e8a7dd9be70cbce269067ccbff858ac8d5a115494b891R558)

LoisRForgeFlow commented:
@janikvonrotz can you have a look at travis?, seems that some test is failing now.

janikvonrotz commented:
Travis log:

```
----------

bad query: INSERT INTO "sale_blanket_order" ("id", "create_uid", "create_date", "write_uid", "write_date", "company_id", "name", "note", "partner_id", "payment_term_id", "pricelist_id", "team_id", "validity_date") VALUES (nextval('sale_blanket_order_id_seq'), 1, (now() at time zone 'UTC'), 1, (now() at time zone 'UTC'), 1, 'Draft', '', 72, 1, 24, 1, '2021-11-30') RETURNING id

ERROR: null value in column "partner_invoice_id" violates not-null constraint

DETAIL:  Failing row contains (1, null, Draft, 72, null, null, 24, null, 1, null, null, null, 2021-11-30, null, , null, 1, 1, null, null, null, null, 1, 2021-12-01 10:38:54.901889, 1, 2021-12-01 10:38:54.901889).

 

ERROR: TestSaleBlanketOrders.test_01_create_blanket_order

Traceback (most recent call last):

  File "/home/travis/build/OCA/sale-workflow/setup/sale_blanket_order/odoo/addons/sale_blanket_order/tests/test_blanket_orders.py", line 82, in test_01_create_blanket_order
```

Probably need to update the test scritpts.

tafaRU commented:
@janikvonrotz all the tests failed during the creation of the `sale.blanket.order` due to:

```psycopg2.IntegrityError: null value in column "partner_invoice_id" violates not-null constraint
DETAIL:  Failing row contains (7, null, Draft, 78, null, null, 30, null, 1, null, null, null, 2022-12-01, null, , null, 1, 1, null, null, null, null, 1, 2021-12-01 10:38:56.644462, 1, 2021-12-01 10:38:56.644462).
```
While the other errors are related to other modules.

BTW, please don't put all the log error in the comment, just put the link, thanks!

Personally I'd prefer move the commits, not strictly related to the migration, into another PR (depending by this) in order to speed up the merge of the current one. What do you think?

janikvonrotz commented:
@tafaRU thanks for helping. I removed parts of the travis log.

With commit 0d31e3b I introduced two new required fields: `partner_invoice_id` and `partner_shipping_id`. 
I assume that these fields are breaking the tests.

As I am relatively new to Odoo development, I am not sure how to run tests for one module locally [^1].

I think updating the tests would be easy:

Change the occurrence of:

```py
"partner_id": self.partner.id,
```

To: 

```py
"partner_id": self.partner.id,
"partner_invoice_id": self.partner.id,
"partner_shipping_id": self.partner.id,
```

Regarding your question to the split the PR, I am not sure if I am able to do that.
Let me check if I can fix the test and  pass Travis.

[^1]: https://github.com/Mint-System/Odoo-Development/blob/e0df70d5245f7e9bdebf27620c7c6fb36f7a12da/task#L212

tafaRU commented:
> Regarding your question to the split the PR, I am not sure if I am able to do that.

If you agree, I can take charge of the module migration, and once the (new) PR will be merged, you can propose a new one by simply cherry-picking* your improvements commits. And only at this point we can worry about how to fix tests.

\* by using `git cherry-pick SHA_commit` i.e: git `cherry-pick 2b86ae79fe46f9f82d9e88188d55311605575214`

janikvonrotz commented:
@tafaRU I agree. Thanks for helping!

janikvonrotz committed:
[FIX] sale_blanket_order: fix tests by adding required fields

tafaRU commented on sale_blanket_order/wizard/create_sale_orders.py:
```diff
@@ -183,7 +183,7 @@ class BlanketOrderWizardLine(models.TransientModel):
 
     wizard_id = fields.Many2one("sale.blanket.order.wizard")
-    blanket_line_id = fields.Many2one("sale.blanket.order.line")
+    blanket_line_id = fields.Many2one("sale.blanket.order.line", store="True")
```
What's the reason for this? The field is stored by default, you don't need to add it the attribute`store`.

janikvonrotz replied:
This is a fix I copied from PR #1608 [^1] after a lot of troubleshooting. If this is not set to `True` the wizard cannot persist the lines when confirming. It will be reset and the wizard lines are empty. I was not able to investigate further, but I assume this is related to some change in v14. Also note that this is a transient model.

[^1]: https://github.com/OCA/sale-workflow/pull/1608/files#diff-2565286591068d712b8e7b02a8c2580442f249c417077cbd482b6c1b3b6403b4

tafaRU commented on sale_blanket_order/wizard/create_sale_orders.xml:
```diff
@@ -8,6 +8,11 @@
                 <group>
                     <field name="line_ids" nolabel="1">
                         <tree create="false" editable="bottom">
+                            <field
```
Also this change: is it needed and why?

janikvonrotz replied:
Refers to my last comment: https://github.com/OCA/sale-workflow/pull/1785/files/6391c73a8f54d6682c6cf828e921bed7cd569d52..83b003ff3a3200a291c35ab04f36f434c08bee35#r773338386

tafaRU commented:
Superseeded by https://github.com/OCA/sale-workflow/pull/1865
